frontends/concrete-python/concrete/fhe/mlir/context.py

```diff
@@ -2200,66 +2231,116 @@ def extract_bits(
         start = bits.start or MIN_EXTRACTABLE_BIT
         stop = bits.stop or (MAX_EXTRACTABLE_BIT if step > 0 else (MIN_EXTRACTABLE_BIT - 1))
 
-        bits_and_their_positions = []
-        for position, bit in enumerate(range(start, stop, step)):
-            bits_and_their_positions.append((bit, position))
+        bits = list(range(start, stop, step))
+
+        bits_and_their_positions = ((bit, position) for position, bit in enumerate(bits))
 
         bits_and_their_positions = sorted(
             bits_and_their_positions,
             key=lambda bit_and_its_position: bit_and_its_position[0],
         )
 
+        # we optimize bulk extract in low precision, used for identity
+        if (
+            len(bits) > 1
+            and LUT_COSTS.get(x.type.bit_width, float("inf")) <= (max(bits) + 1) * LUT_COSTS[1]
+        ):
+
+            def is_positive(v):
+                return x.type.is_unsigned or v < 2 ** (x.type.bit_width - 1)
+
+            def to_signed(v):
+                if is_positive(v):
+                    return v
+                return -(2 ** (x.type.bit_width) - v)
+
+            table = [
+                sum(
+                    ((to_signed(v) >> bit) & 1) << position
+                    for bit, position in bits_and_their_positions
+                )
+                + (0 if is_positive(v) else 2 ** (resulting_type.bit_width + 1))
+                for v in range(2**x.type.bit_width)
+            ]
+            tlu_result = self.tlu(resulting_type, x, table)
+            return self.to_signedness(tlu_result, of=resulting_type)
+
+        def same_type_as(type_, bit_width):
```

Comment:
I was thinking of `type.fork(bit_width=10)` or similar for that, maybe you can do it if it's easy? Not so important though.

Reply:
I don't see how, I need the context object to do it. It relies on the context.typeof function.
but could be context.type_fork(type_, bit_width) since it's probably useful in other places in context.

Reply:
I'd go for `context.fork_type(original_type, **overrides)` yes!